encoding/test_encoding.py

```diff
@@ -0,0 +1,242 @@
+from framework import tester
+from helpers import deproxy
+
+__author__ = "Tempesta Technologies, Inc."
+__copyright__ = "Copyright (C) 2022 Tempesta Technologies, Inc."
+__license__ = "GPL2"
+
+
+DATE = deproxy.HttpMessage.date_time_string()
+
+
+class CommonUtils:
+    def start_all(self):
+        srv = self.get_server("backend")
+        #        srv.keep_alive = 1
+        srv.start()
+        self.start_tempesta()
+        self.start_all_clients()
+        self.deproxy_manager.start()
+        self.assertTrue(srv.wait_for_connections(1))
+
+    def send_req(self, client):
+        request = (
+            "GET / HTTP/1.1\r\n"
+            "Host: localhost\r\n"
+            "Accept-Encoding: gzip, br, chunked\r\n"
+            "Connection: keep-alive\r\n"
+            "\r\n"
+        )
+        client.make_requests(request)
+        got_response = client.wait_for_response(timeout=5)
+        return got_response
+
+
+class TestChunkedIsNotLast(tester.TempestaTest, CommonUtils):
+    """
+    Ensures that if `chunked` is not the last encoding listed
+    in `Transfer-Encoding` header, the connection is dropped.
+    """
+
+    clients = [
+        {"id": "client", "type": "deproxy", "addr": "${tempesta_ip}", "port": "80"}
+    ]
+    backends = [
+        {
+            "id": "backend",
+            "type": "deproxy",
+            "port": "8000",
+            "response": "static",
+            "response_content":
+            "HTTP/1.1 200 OK\r\n"
+            "Connection: keep-alive\r\n"
+            "Content-type: text/html\r\n"
+            f"Last-Modified: {DATE}\r\n"
+            f"Date: {DATE}\r\n"
+            "Server: Deproxy Server\r\n"
+            "Transfer-Encoding: chunked, gzip\r\n\r\n"
+            "the body does not actually matter",
+        }
+    ]
+    tempesta = {
+        "config": """
+        server ${server_ip}:8000;
+        """
+    }
+
+    def test(self):
+        self.start_all()
+
+        client = self.get_client("client")
+        server = self.get_server("backend")
+
+        got_response = self.send_req(client)
+        self.assertTrue(
+            client.connection_is_closed(), "The connection has to be closed"
+        )
+        self.assertFalse(got_response, "There should be no response")
```

Comment:
In the test `TestChunkedIsNotLast` backend(not tempesta) must close connection after sending all data and must not send `Content-Length`. Tempesta adds `Content-Length` and forwards response. Test has to verify received from tempesta `Content-Length`.

Reply:
With the latest changes to the tests I get:
```
test (encoding.test_encoding.TestChunkedIsNotLast) ... FAIL

======================================================================
FAIL: test (encoding.test_encoding.TestChunkedIsNotLast)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/tempesta-test/encoding/test_encoding.py", line 77, in test
    self.assertTrue(got_response, "There should be a response")
AssertionError: False is not true : There should be a response

----------------------------------------------------------------------
Ran 1 test in 8.385s

```
and interestingly enough, if I add "Connection: close" header to the deproxy static response, I get this:
```
[Mon Oct 10 13:07:04 2022] [tempesta fw] Warning: Parser error: state=Resp_HdrConnectionV input(-8)=0x63('ection: close
') data_len=261 off=185    Transfer-Encoding: chunked, gzip
[Mon Oct 10 13:07:04 2022] [tempesta fw] Warning: response dropped: processing error: 127.0.0.1
```
and then do I get a response, but it's a `502`. Even though now the response has "Content-Length: 0" and the test passes, this is not what we're expecting here, right?

Reply:
"backend must close connection" implies backend must close TCP connection immediately after sending data. You can use deproxy param [`keep_alive`](https://github.com/tempesta-tech/tempesta-test/blob/master/framework/deproxy_server.py#L61) to achieve such behaviour. `Connection: close` doesn't matter. Error that appears after adding `Connection: close` is likely due to doubling of header.